### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive when used with a any type coming from a generic

### ISSUE THREAD

**Issue by julienw** [typescript-eslint/typescript-eslint#11298]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=FAEwpgxgNghgTmABAMwK4DsIBcCWB7dRVAZzABk8Zw4ARGLGAHgBVEBeRGdATwD4AKAJQAuRMwDcwYBALEsiAN4o8eRAF92RUhSpha9GEM7FFyvKLlwc6AObrJQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RIAGnBgAviD1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare function useLoaderData<T = any>(): T;

const { foo } = useLoaderData() as { foo: string };
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error.

### Actual Result

False positive.

### Additional Info

I use `useLoaderData` from the `react-router` package.

I could rewrite the code like this:
```js
declare function useLoaderData<T = any>(): T;

const { foo } = useLoaderData<{ foo: string }>();
```
But this is functionally equivalent, and I think the rule shouldn't error in this case.

**Comment by bradzacher** [comment]
duplicate of #10722
